paramiko/auth_handler.py

```diff
@@ -243,13 +353,43 @@ def _parse_service_accept(self, m):
                 password = bytestring(self.password)
                 m.add_string(password)
             elif self.auth_method == 'publickey':
-                m.add_boolean(True)
-                m.add_string(self.private_key.get_name())
-                m.add_string(self.private_key)
-                blob = self._get_session_blob(
-                    self.private_key, 'ssh-connection', self.username)
-                sig = self.private_key.sign_ssh_data(blob)
-                m.add_string(sig)
+                if self.pkcs11pin is None:
+                    # Private Key
+                    m.add_boolean(True)
+                    m.add_string(self.private_key.get_name())
+                    m.add_string(self.private_key)
+                    blob = self._get_session_blob(
+                        self.private_key, 'ssh-connection', self.username)
+                    sig = self.private_key.sign_ssh_data(blob)
+                    m.add_string(sig)
+                else:
+                    # Smartcard PKCS11 Private Key
+                    keym = Message()
+                    key_name = ''
+                    pkcs11_tmp_pub = load_ssh_public_key(
+                        self._pkcs11_get_public_key(),
+                        cryptography.hazmat.backends.default_backend())
+                    if isinstance(pkcs11_tmp_pub, RSAPublicKey):
+                        key_name = 'ssh-rsa'
+                    elif isinstance(pkcs11_tmp_pub, DSAPublicKey):
+                        key_name = 'ssh-dss'
+                    elif isinstance(pkcs11_tmp_pub, EllipticCurvePublicKey):
+                        key_name = 'ssh-ecdsa'
+                    else:
+                        raise Exception("Unable to determine pkcs11 key type")
+                    keym.add_string(key_name)
+                    keym.add_mpint(pkcs11_tmp_pub.public_numbers().e)
+                    keym.add_mpint(pkcs11_tmp_pub.public_numbers().n)
```

Comment:
If key types besides ssh-rsa are possible, it is a bit more complicated to support them - they do not have "e" and "n", they have different sets of defining values.

Paramiko's `RSAKey` has an `asbytes()` method which does exactly this. The other key types also have that method, to pack the appropriate Message for that type. So the way to do this would be similar to how `HostKeyEntry.from_line()` does it: https://github.com/paramiko/paramiko/blob/2.2.1/paramiko/hostkeys.py#L325